code/WorkInProgress/telescope/telescopeMisc.dm

```diff
@@ -122,19 +101,44 @@ var/list/magnet_locations = list()
 			return 1
 		return 0
 
-	Topic(href, href_list)
-		if (src.busy)
-			return
-
-		if (get_dist(usr, src) > 1 || usr.z != src.z)
-			return
+/obj/machinery/lrteleporter/ui_interact(mob/user, datum/tgui/ui)
+	ui = tgui_process.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "LongRangeTeleporter", name)
+		ui.open()
+
+/obj/machinery/lrteleporter/ui_data(mob/user)
+	var/list/destinations = list()
+	for(var/A in special_places)
+		destinations += list(list(
+			"destination" = "[A]",
+			"ref" = null))
+
+	. = list(
+		"destinations" = destinations
+	)
+
+/obj/machinery/lrteleporter/ui_static_data(mob/user)
+	. = list(
+		"send_allowed" = TRUE,
+		"receive_allowed" = TRUE,
+		"syndicate" = FALSE
+	)
+
+/obj/machinery/lrteleporter/ui_act(action, params)
+	. = ..()
+	if(.)
+		return
+	if(busy) return
+	if(!in_interact_range(src, usr) || usr.z != src.z) return
```

Comment:
TL;DR at bottom of comment.

This *should* be covered by `ui_act` on `datum` (see [here](https://github.com/goonstation/goonstation/blob/7f51873c5dbe6b80d3567f1b8121bc16729a18ce/code/modules/tgui/external.dm#L78)).

It returns `1` if the `ui.status` is not `UI_INTERACTIVE`, which is why the `if (.) return` bit is done above.

`ui.status` is generated from the `ui_status` proc, which exists on `/obj/machinery` [here](https://github.com/goonstation/goonstation/blob/7f51873c5dbe6b80d3567f1b8121bc16729a18ce/code/obj/machinery.dm#L186). The `state` it uses is determined the `ui_state` proc, which exists on `/obj/machinery` [here](https://github.com/goonstation/goonstation/blob/7f51873c5dbe6b80d3567f1b8121bc16729a18ce/code/obj/machinery.dm#L180).

In your case, it doesn't require physical access, so the `tgui_default_state` is used. That's defined [here](https://github.com/goonstation/goonstation/blob/7f51873c5dbe6b80d3567f1b8121bc16729a18ce/code/modules/tgui/states/default.dm).

The assorted stuff in there determines whether the UI should close, be open but not updating, open and updating but not allowing inputs, or be fully interactive (UI_INTERACTIVE).

Each type of mob handles that a little differently, and that's sort of the point (i.e. whether a human should be able to act is different from whether the AI should be able to).

Specifically for humans, it's only interactive if:
* they have a client connected, they are conscious, they are not unable to act for a few reasons (e.g. cuffs) - all of this is covered by `shared_ui_interaction`.
* the thing is not in a storage item, has the result of `get_dist` be `<= 1` (this does *not* deal with the z-level bit), the thing is in view (this *should* cover the z-level bit, I think?), is in interact range (uses `in_interact_range`) - all of this is covered by `shared_living_ui_distance`.

It's a little different for cyborgs/AI, but the key is that "you don't have to worry about it, it's covered by this stuff".

It *would* be worth sanity checking that the `in view` bit of [this](https://github.com/goonstation/goonstation/blob/7f51873c5dbe6b80d3567f1b8121bc16729a18ce/code/modules/tgui/states.dm#L117) does account for z-levels. If it does, you can remove the line this comment is attached to. If it does not, we should think about solving that.

I appreciate that this is an in-depth comment, meant to educate about the whole thing. TL;DR: this line *shouldn't* be needed, if it is then we should change something else instead.

Reply:
Added #5683 to capture desired behavior.